### DISCUSSION TITLE
Forwarding signature to decorated functions

### DISCUSSION THREAD

**Discussion by pawamoy** [microsoft/pyright#4930] [Q&A]
I'm trying to annotate my decorators so that the decorated functions get the same prompts when typing code in VSCode.

I've found useful information in https://github.com/microsoft/pylance-release/issues/125, https://github.com/microsoft/pylance-release/issues/1470, and https://github.com/microsoft/pyright/issues/774.

I'm able to make it work, but I do have two question. First, here's my code:

```python
from __future__ import annotations

import sys
from functools import wraps
from typing import Callable, TypeVar

if sys.version_info < (3, 10):
    from typing_extensions import ParamSpec
else:
    from typing import ParamSpec

_P = ParamSpec("_P")
_R = TypeVar("_R")


class LazyCallable:
    def __init__(self, call: Callable[_P, _R], args: tuple, kwargs: dict) -> None:
        self.call = call
        self.args = args
        self.kwargs = kwargs

    def __call__(self) -> Any:  # noqa: D102
        return self.call(*self.args, **self.kwargs)


def lazy(call: Callable[_P, _R]) -> Callable[_P, LazyCallable]:
    @wraps(call)
    def lazy_caller(*args: _P.args, **kwargs: _P.kwargs) -> LazyCallable:
        return LazyCallable(call, args, kwargs)
    return lazy_caller
```

With this code I can lazify a function like this:

```python

@lazy
def greet(name):
    print(f"Hello {name}!")

# or
def greet(name):
    print(f"Hello {name}!")

greet = lazy(greet)
```

Then when I start typing `greet(`, the IDE shows me the parameters of the original `greet` function :+1:

---

My questions now:

1. if I declare `_C = Callable[_P, _R]` and replace occurrences of `Callable[_P, _R]` with it, it stops working (i.e. VSCode will show `(...) -> Any`). Is this expected?

```python
_P = ParamSpec("_P")
_R = TypeVar("_R")
_C = Callable[_P, _R]


class LazyCallable:
    def __init__(self, call: _C, args: tuple, kwargs: dict) -> None:
        self.call = call
        self.args = args
        self.kwargs = kwargs

    def __call__(self) -> Any:  # noqa: D102
        return self.call(*self.args, **self.kwargs)


def lazy(call: _C) -> Callable[_P, LazyCallable]:
    @wraps(call)
    def lazy_caller(*args: _P.args, **kwargs: _P.kwargs) -> LazyCallable:
        return LazyCallable(call, args, kwargs)
    return lazy_caller
```

2. how should I annotate `LazyCallable.__call__` to say that it will return the same thing as the given `call` function? Mypy complains if I use `-> _R`:
    - *"A function returning TypeVar should receive at least one argument containing the same TypeVar"*
    - *"Incompatible return value type (got "_R", expected "_R")"*

    Is this a use-case for `typing.Protocol`?

**Comment by erictraut** [comment]
In your second example, you are defining a generic type alias called `_C`. This type alias has two type parameters — a ParamSpec called `_P` and a TypeVar called `_R`. When you use this type alias, you need to supply two corresponding type arguments, but you've failed to supply these arguments when you use `_C` later in your code. Since you omitted these type arguments, type checkers will assume that you mean `_C[..., Any]`. 

I recommend enabling the `reportMissingTypeArgument` check in pyright. It will then tell you when you've omitted type arguments for a generic class or type alias. You can do this globally for your project using the config file or you enable it within a single file by adding the following comment at the top of the file:

```python
# pyright: reportMissingTypeArgument=true
```

The other problem with the second example is that the `LazyCallable` class should specify that `_P` and `_R` are type parameters for the class. 

Here's updated code with these fixes applied:

```python
class LazyCallable(Generic[_P, _R]):
    def __init__(self, call: _C[_P, _R], *args: _P.args, **kwargs: _P.kwargs) -> None:
        self.call = call
        self.args = args
        self.kwargs = kwargs

    def __call__(self) -> Any:
        return self.call(*self.args, **self.kwargs)


def lazy(call: _C[_P, _R]) -> Callable[_P, LazyCallable[_P, _R]]:
    @wraps(call)
    def lazy_caller(*args: _P.args, **kwargs: _P.kwargs) -> LazyCallable[_P, _R]:
        return LazyCallable(call, *args, **kwargs)

    return lazy_caller
```

**Reply by pawamoy** [reply]
Thank you so much @erictraut, this is really helpful. Despite my years of Python, I'm still a newbie when it comes to the typing system. I'll try what you suggest and report back!

**Reply by erictraut** [reply]
Type variable scoping rules in Python are very confusing today. They trip up many developers, so you are not alone. 

I've submitted [PEP 695](https://peps.python.org/pep-0695/) to improve this issue. I'm happy to report that the PEP was just accepted by the Python steering committee for inclusion in Python 3.12.

**Reply by pawamoy** [reply]
Awesome :tada: Congrats on your accepted PEP! Looking forward to easier type annotations :smile:
